credoai/modules/model_assessments/nlp_generator.py

```diff
@@ -0,0 +1,496 @@
+                return response
+            else:
+                return "nlp generator error"
+        except:  # no response
+            return "nlp generator error"
+
+    def run(self, n_iterations=1, verbose=1):
+        """Run the generations and assessments
+
+        Parameters
+        ----------
+        n_iterations : int, optional
+            Number of times to generate responses for a prompt, by default 1
+            Increase if your generation model is stochastic for a higher confidence
+        verbose : int, optional
+            Progress messages will be printed if 1, by default 1
+
+        Returns
+        -------
+        pandas.dataframe
+            Complete assessment results for all the prompts and iterations
+            Schema: ['group', 'subgroup', 'prompt', 'run', 'response', 'value', 'assessment_attribute']
+
+        Raises
+        ------
+        Exception
+            Occurs if 'api_key' not provided in assessment_config and it is needed
+        Exception
+            Occurs if assessment_config is not valid
+        """
+        df = self._get_prompts(self.assessment_config["prompts_dataset"])
+        if verbose == 1:
+            logging.info(
+                "Loaded the prompts dataset "
+                + self.assessment_config["prompts_dataset"]
+            )
+
+        # Perform prerun checks
+        self.perform_prerun_checks()
+
+        # Generate and record responses for the prompts with all the generation models n_iterations times
+        dfruns = pd.DataFrame(
+            columns=["generation_model", "group", "subgroup", "prompt", "run"]
+        )
+
+        for gen_name, gen_fun in self.generation_config.items():
+            for i in range(n_iterations):
+                if verbose == 1:
+                    logging.info(
+                        "Performing Generation Iteration "
+                        + str(i + 1)
+                        + " of "
+                        + str(n_iterations)
+                        + " with Generation Model "
+                        + gen_name
+                    )
+                df["response"] = df["prompt"].apply(
+                    lambda x: self._gen_fun_robust(x, gen_fun)
+                )
+                df["run"] = i + 1
+                df["generation_model"] = gen_name
+                dfruns = dfruns.append(df)
+
+        # Assess the responses for the input assessment attributes
+        if verbose == 1:
+            logging.info("Performing assessment of the generated responses")
+
+        dfrunst = dfruns[
+            dfruns["response"] != "nlp generator error"
+        ].copy()  # exclude cases where generator failed to generate a response
+
+        dfrunst_assess = pd.DataFrame(
+            data=None, columns=list(dfrunst.columns) + ["value", "assessment_attribute"]
+        )
+
+        if "assess_with_builtin_models" not in self.assessment_config:
```

Comment:
why not provide the functionality to both use a build-in-model AND there own functions? Wouldn't they all just end up being different assessments of the resultant generators?

Reply:
now enabled. `assessment_functions` parameter can take both types simultaneously